File: web/src/main/java/com/personal/image_toolkit/ImageCrawler.java
package com.example.image_toolkit;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Duration;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;

/**
 * Downloads all images from a webpage using Selenium and a Java HTTP client.
 * Implements the WebCrawler abstract class.
 */
public class ImageCrawler extends WebCrawler {

    private final String targetUrl;
    private final int skipFirst;
    private final int skipLast;
    private final ImageCrawlerListener listener;

    public ImageCrawler(String url, int skipFirst, int skipLast, boolean headless, String downloadDir, String screenshotDir, String browser, ImageCrawlerListener listener) {
        super(headless, downloadDir, screenshotDir, browser);
        this.targetUrl = url;
        this.skipFirst = skipFirst;
        this.skipLast = skipLast;
        this.listener = listener;
        System.out.printf("ImageCrawler ready: %s%n", url);
    }

    @Override
    public boolean login(Map<String, String> credentials) {
        listener.onStatus("No login needed.");
        return true;
    }
    
    // Helper to get a unique filename like: cat.jpg -> cat (1).jpg
    private Path getUniqueFilename(Path filepath) {
        Path parent = filepath.getParent();
        String filename = filepath.getFileName().toString();
        int dotIndex = filename.lastIndexOf('.');
        String base = (dotIndex == -1) ? filename : filename.substring(0, dotIndex);
        String ext = (dotIndex == -1) ? "" : filename.substring(dotIndex);
        
        int counter = 1;
        Path newPath = filepath;
        while (Files.exists(newPath)) {
            newPath = parent.resolve(String.format("%s (%d)%s", base, counter, ext));
            counter++;
        }
        return newPath;
    }

    @Override
    public void processData() {
        listener.onStatus("Loading page...");
        if (!navigateToUrl(targetUrl, false)) {
            listener.onStatus("Failed to load page.");
            return;
        }

        waitBeforePageLoad(10, new String[]{}, null);
        listener.onStatus("Scanning for images...");

        try {
            List<WebElement> images = driver.findElements(By.tagName("img"));
            int totalFound = images.size();
            int skipTotal = skipFirst + skipLast;
            
            if (skipTotal >= totalFound) {
                listener.onStatus("Not enough images to skip.");
                return;
            }

            // Sublist equivalent to Python slice images[skipFirst:-skipLast]
            List<WebElement> imagesToProcess = images.subList(skipFirst, totalFound - skipLast);
            int totalImagesToDownload = imagesToProcess.size();

            listener.onProgress(0, totalImagesToDownload);
            listener.onStatus(String.format("Found %d images. Skipping %d. Downloading %d...", 
                                            totalFound, skipTotal, totalImagesToDownload));

            Set<String> uniqueUrls = imagesToProcess.stream()
                .map(img -> img.getAttribute("src"))
                .filter(src -> src != null && !src.startsWith("data:"))
                .map(src -> {
                    try {
                        return new URL(new URL(driver.getCurrentUrl()), src).toExternalForm();
                    } catch (Exception e) {
                        return null; // Ignore invalid URLs
                    }
                })
                .filter(url -> url != null)
                .collect(Collectors.toSet());

            listener.onStatus(String.format("Downloading %d unique images...", uniqueUrls.size()));
            
            AtomicLong downloadedCount = new AtomicLong(0);

            for (String url : uniqueUrls) {
                if (_downloadImageFromUrl(url)) {
                    downloadedCount.incrementAndGet();
                    listener.onProgress((int) downloadedCount.get(), uniqueUrls.size());
                }
                TimeUnit.MILLISECONDS.sleep(100);
            }

            listener.onStatus(String.format("Downloaded %d images! Skipped %d.", downloadedCount.get(), skipTotal));

        } catch (Exception e) {
            listener.onStatus("Error: " + e.getMessage());
        }
    }

    private boolean _downloadImageFromUrl(String url) {
        try {
            URI uri = new URI(url);
            String path = uri.getPath();
            String filename = path.substring(path.lastIndexOf('/') + 1).split("\\?")[0];
            
            if (filename.isEmpty() || !filename.contains(".")) {
                filename = String.format("image_%d.jpg", System.currentTimeMillis());
            }

            Path savePath = downloadDir.resolve(filename);
            savePath = getUniqueFilename(savePath);

            HttpClient client = HttpClient.newBuilder()
                    .connectTimeout(Duration.ofSeconds(15))
                    .build();

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(uri)
                    .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64)")
                    .GET()
                    .build();

            HttpResponse<InputStream> response = client.send(request, HttpResponse.BodyHandlers.ofInputStream());

            if (response.statusCode() == 200) {
                try (InputStream is = response.body()) {
                    Files.copy(is, savePath);
                }
                listener.onImageSaved(savePath.toString());
                return true;
            } else {
                throw new IOException("HTTP error code: " + response.statusCode());
            }

        } catch (Exception e) {
            System.err.printf("Failed %s: %s%n", url, e.getMessage());
            return false;
        }
    }

    public int run() {
        try {
            login(null);
            processData();
            return 0;
        } finally {
            close();
        }
    }
}

File: desktop/src/web/ImageCrawler.py
import os
import time
import requests

from urllib.parse import urlparse, urljoin
from selenium.webdriver.common.by import By
from PySide6.QtCore import QObject, Signal
from .Crawler import WebCrawler


class QtABCMeta(type(QObject), type(WebCrawler)):
    """Combines Qt's QObject metaclass with ABC's metaclass."""
    pass


class ImageCrawler(WebCrawler, QObject, metaclass=QtABCMeta):
    """Downloads all images from a webpage with live Qt signals."""
    
    # === SIGNALS ===
    on_progress = Signal(int, int)      # (current, total)
    on_status = Signal(str)            # status message
    on_image_saved = Signal(str)       # saved file path

    def __init__(self, url, skip_first=0, skip_last=9, headless=False, download_dir=None, screenshot_dir=None, browser="brave"):
        QObject.__init__(self)         # ← Initialize QObject
        WebCrawler.__init__(self, headless, download_dir, screenshot_dir, browser)
        self.target_url = url
        self.skip_first = skip_first
        self.skip_last = skip_last
        print(f"ImageCrawler ready: {url}")

    def login(self, credentials=None):
        self.on_status.emit("No login needed.")
        return True
    
    def get_unique_filename(self, filepath):
        """Returns a unique filename like: cat.jpg → cat (1).jpg"""
        base, ext = os.path.splitext(filepath)
        counter = 1
        new_path = filepath
        while os.path.exists(new_path):
            new_path = f"{base} ({counter}){ext}"
            counter += 1
        return new_path

    def process_data(self): # 4 12
        self.on_status.emit("Loading page...")
        if not self.navigate_to_url(self.target_url, take_screenshot=False):
            self.on_status.emit("Failed to load page.")
            return

        self.wait_for_page_to_load(timeout=10)
        self.on_status.emit("Scanning for images...")

        try:
            images = self.driver.find_elements(By.TAG_NAME, "img")
            total_found = len(images)
            skip_total = self.skip_first + self.skip_last
            if skip_total >= total_found:
                self.on_status.emit("Not enough images to skip.")
                return

            images_to_process = images[self.skip_first:-self.skip_last]  # ← SKIP LAST 5
            total = len(images_to_process)

            self.on_progress.emit(0, total)
            self.on_status.emit(f"Found {total_found} images. Skipping {skip_total}. Downloading {total}...")

            urls = {
                urljoin(self.driver.current_url, img.get_attribute("src"))
                for img in images_to_process
                if img.get_attribute("src") and not img.get_attribute("src").startswith("data:")
            }

            self.on_status.emit(f"Downloading {len(urls)} unique images...")

            for i, url in enumerate(urls):
                if self._download_image_from_url(url):
                    self.on_progress.emit(i + 1, len(urls))
                time.sleep(0.1)

            self.on_status.emit(f"Downloaded {len(urls)} images! Skipped {skip_total}.")

        except Exception as e:
            self.on_status.emit(f"Error: {e}")

    def _download_image_from_url(self, url):
        try:
            parsed = urlparse(url)
            filename = os.path.basename(parsed.path).split('?')[0]
            if not filename or '.' not in filename:
                filename = f"image_{int(time.time() * 1000)}.jpg"

            save_path = os.path.join(self.download_dir, filename)
            save_path = self.get_unique_filename(save_path)  # ← NEVER OVERWRITE

            response = requests.get(url, stream=True, timeout=15, headers={
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'
            })
            response.raise_for_status()

            with open(save_path, 'wb') as f:
                for chunk in response.iter_content(8192):
                    f.write(chunk)

            self.on_image_saved.emit(save_path)
            return True

        except Exception as e:
            print(f"Failed {url}: {e}")
            return False

    def run(self):
        try:
            self.login()
            self.process_data()
        finally:
            self.close()
        return 0
